benchmarks/src/main/kotlin/bfschannel/BfsChannelBenchmark.kt

```diff
@@ -0,0 +1,362 @@
+    private val _neighbours = arrayListOf<Node>()
+    val neighbours: List<Node> = _neighbours
+
+    fun addNeighbour(node: Node) {
+        _neighbours.add(node)
+    }
+
+    val distance = AtomicLong(Long.MAX_VALUE)
+}
+
+/**
+ * This channel implementation does not suspend on sends and closes itself if the number of waiting receivers exceeds [maxWaitingReceivers].
+ */
+@Suppress("INVISIBLE_REFERENCE", "INVISIBLE_MEMBER", "SubscriberImplementation")
+internal class SelfClosingChannel<E>(private val maxWaitingReceivers: Int) : LinkedListChannel<E>() {
+    private val waitingReceivers = AtomicLong(0)
+
+    @Suppress("CANNOT_OVERRIDE_INVISIBLE_MEMBER")
+    override fun onReceiveEnqueued() {
+        val waitingReceivers = waitingReceivers.incrementAndGet()
+        if (waitingReceivers >= maxWaitingReceivers) close()
+    }
+
+    @Suppress("CANNOT_OVERRIDE_INVISIBLE_MEMBER")
+    override fun onReceiveDequeued() {
+        waitingReceivers.decrementAndGet()
+    }
+}
+
+abstract class GraphCreator(val name: String) {
+    fun getGraph(): List<Node> {
+        if (!Paths.get(graphFileName).toFile().exists()) {
+            generateOrDownloadGraphFile(graphFileName)
+        }
+        return parseGraphFile(graphFileName)
+    }
+
+    abstract val graphFileName: String
+
+    abstract fun generateOrDownloadGraphFile(graphFileName: String)
+
+    abstract fun parseGraphFile(graphFileName: String): List<Node>
+}
+
+class RandomGraphCreator(name: String, private val nodes: Int, private val edges: Int) : GraphCreator(name) {
+    override fun generateOrDownloadGraphFile(graphFileName: String) {
+        println("Generating $graphFileName as a random graph with $nodes nodes and $edges edges")
+        val graphNodes = randomConnectedGraph(nodes, edges)
+        writeGeneratedGraphToGrFile(graphFileName, graphNodes)
+        println("Generated $graphFileName")
+    }
+
+    override val graphFileName: String
+        get() = "out/$name.gr"
+
+    override fun parseGraphFile(graphFileName: String): List<Node> = parseGrFile(graphFileName, false)
+}
+
+class DownloadingGraphCreator(name: String, private val type: GraphType, private val url: String) : GraphCreator(name) {
+    override val graphFileName: String
+        get() = "out/$name.${type.ext()}.${if (type.gzipped()) "gz" else ""}"
+
+    override fun generateOrDownloadGraphFile(graphFileName: String) {
+        println("Downloading $graphFileName from $url")
+        val input = Channels.newChannel(URL(url).openStream())
+        val output = FileOutputStream(graphFileName)
+        output.channel.transferFrom(input, 0, Long.MAX_VALUE)
+        println("Downloaded $graphFileName")
+    }
+
+    override fun parseGraphFile(graphFileName: String): List<Node> {
+        return when (type) {
+            GraphType.GR_GZ -> parseGrFile(graphFileName, type.gzipped())
+            GraphType.TXT_GZ -> parseTxtFile(graphFileName, type.gzipped())
+        }
+    }
+}
+
+enum class GraphType {
```

Comment:
You do not need this enum. It is better to have a common abstract class for all the graph types, so that each graph is represented with an instance of this class.